### PR TITLE
Add dropbear as an ssh server

### PR DESCRIPTION
**Pull request by onde2rock**
Edit: Finished.

- [x] SDL linux : tested and working
- [x] Kindle5 : tested and working
- [ ] recent Kindle : looking for tester, [zip](https://github.com/onde2rock/koreader/releases/download/test/koreader-kindle-arm-linux-gnueabi-v2015.11-1626-gfdd57fb_2018-04-15.zip)
- [x] Kobo : looking for tester, [zip](https://github.com/onde2rock/koreader/releases/download/test/koreader-kobo-arm-linux-gnueabihf-v2015.11-1626-gfdd57fb_2018-04-15.zip)
- [ ] Pocketbook : looking for tester 
- [x] Android : not supported
- [x] Windows : not supported

### CODE REVIEW

File: Makefile.third [hunk 1]
```diff
@@ -317,6 +317,25 @@ ifdef ANDROID
 	readelf -d $@ | grep "no dynamic" \
 		|| echo "warning: Android L+ support only PIE binary and can't run this"
 endif
+# ===========================================================================
+# dropbear
+
+$(OUTPUT_DIR)/dropbear: $(THIRDPARTY_DIR)/dropbear/CMakeLists.txt
```
Comment (Frenzie) [inline Makefile.third]: This depends on zlib, so something like:

```
$(OUTPUT_DIR)/dropbear: $(ZLIB) $(THIRDPARTY_DIR)/dropbear/CMakeLists.txt

File: Makefile.third [hunk 2]
```diff
@@ -317,6 +317,25 @@ ifdef ANDROID
...
+	install -d $(DROPBEAR_BUILD_DIR)
+	cd $(DROPBEAR_BUILD_DIR) && \
+		$(CMAKE) -DCC="$(CC)" \
+		-DLIBS="$(if $(or $(ANDROID),$(WIN32),$(DARWIN)),,-lrt) $(if $(ANDROID),-static,)" \
+		$(if $(LEGACY),-DDISABLE_LARGEFILE:BOOL=ON -DDISABLE_FORTIFY:BOOL=ON,) \
```
Comment (Frenzie) [inline Makefile.third]: What's this LEGACY stuff exactly? Is it specifically for the Kindle 4NT (and similarly old devices)?

File: Makefile.third [hunk 3]
```diff
@@ -317,6 +317,25 @@ ifdef ANDROID
...
 		|| echo "warning: Android L+ support only PIE binary and can't run this"
 endif
+# ===========================================================================
+# dropbear
+
+$(OUTPUT_DIR)/dropbear: $(THIRDPARTY_DIR)/dropbear/CMakeLists.txt
+	install -d $(DROPBEAR_BUILD_DIR)
+	cd $(DROPBEAR_BUILD_DIR) && \
+		$(CMAKE) -DCC="$(CC)" \
+		-DLIBS="$(if $(or $(ANDROID),$(WIN32),$(DARWIN)),,-lrt) $(if $(ANDROID),-static,)" \
```
Comment (Frenzie) [inline Makefile.third]: We also want something like

```
-DCPPFLAGS="$(CFLAGS) -I$(ZLIB_DIR)"
```

and

```
-DLDFLAGS="$(LDFLAGS) -L$(ZLIB_DIR)"
```

Or in some kind of LIBS form. I don't know what dropbear wants precisely.

File: Makefile.third [hunk 4]
```diff
@@ -317,6 +317,25 @@ ifdef ANDROID
...
+		$(if $(LEGACY),-DDISABLE_LARGEFILE:BOOL=ON -DDISABLE_FORTIFY:BOOL=ON,) \
+		-DCHOST="$(if $(EMULATE_READER),,$(CHOST))" \
+		$(CURDIR)/$(THIRDPARTY_DIR)/dropbear && \
+		$(MAKE)
+	cp $(DROPBEAR_DIR)/dropbear $(OUTPUT_DIR)/
+	cp $(DROPBEAR_DIR)/dropbearkey $(OUTPUT_DIR)/
+ifdef ANDROID
```
[the remark this answers is not in the document]
Comment (Frenzie) [reply Makefile.third]: What system do you use? Waiting for Travis to show bugs in the Kobo/Kindle compilation without the ability to even try it out is… better than nothing, but pretty terrible. :-)

**Comment by onde2rock** [comment]
@Frenzie Do you know which target do you use to build the release "koreader-kindle-arm-linux-gnueabi" ?
I'm trying to investigate something with the toolchains.

https://github.com/koreader/koreader/blob/5aacc761e1305d8e33accce0b975fc0e8e968454/kodev#L167
https://github.com/koreader/koreader-base/blob/2ed403b617cdb189d47b8eec9700144ec887a748/Makefile.defs#L35-L47

File: Makefile.third (hunk 2 above)
Comment (onde2rock) [reply Makefile.third]: I'm investating and I think it's for device <= kindle3 and pocketbook

**Comment by Frenzie** [comment]
If you mean the nightly build, that one uses `kindle` and `kindle-legacy`.

Also see https://github.com/koreader/koreader/blob/5aacc761e1305d8e33accce0b975fc0e8e968454/kodev#L112-L115

If you mean the Travis build here on GH, that one only tests `TARGET=kindle`

https://github.com/koreader/koreader-base/blob/3f0084f1d7457941303a29c576268944c47df071/.ci/build_script.sh#L21

**Comment by Frenzie** [comment]
> sudo docker run -t -v "${HOME}/.ccache:/home/ko/.ccache" -v "$(pwd):/home/ko//base" "houqp/kokindle:0.0.2"

Oh, are we still using the 0.0.2 image on Travis? For the nightly build we use 0.0.3. Someone should update that.[1]

Btw, you can also build your own: https://github.com/koreader/virdevenv/tree/c3481c22f56d7b66151d5120b3755f261a2c5a3f/docker/ubuntu

I have a baseimage updated to 16.04 (and possibly gcc 5.4) basically ready to go but I'm waiting for a second opinion on that. (Currently the nightly builds run on 14.04 and gcc 4.8).

[1] For my own reference:
* houqp/kokobo:0.0.3
* houqp/kokindle:0.0.3
* houqp/kopb:0.0.1
* houqp/kout:0.0.1
* houqp/kokindle:0.0.2 (huh, the legacy kindle builds on an older image? probably forgot to update that…)
* houqp/koandroid:0.0.6

**Comment by Frenzie** [comment]
@onde2rock The new legacy Kindle build can be found under job artifacts here: https://gitlab.com/koreader/nightly-builds/-/jobs/61975353

It's also available at http://build.koreader.rocks/download/nightly/ but at least for me the local node isn't giving me http://build.koreader.rocks/download/nightly/v2015.11-1643-g5aacc76_2018-04-09/ yet.

**Comment by onde2rock** [comment]
Oh that's awesome !

Also, the kindle4 is in fact the same as the kindle5, so I should'nt have been running the legacy.

So I tried to compile the Kindle5 target and in the meantime, I tested the normal kindle version and it work great! I didn't test all the functionality, but the UI is completely working.
That's nice since this version doesn't need the custom toolchain.

**Comment by Frenzie** [comment]
So that line that kindle-legacy is for "For kindle2/3/4/DXG" is wrong or specifically that 4NT model?

**Comment by onde2rock** [comment]
Quoting https://github.com/koreader/koxtoolchain/blob/47acb959b6169b5d1e546814d1747a8f88484b1b/refs/x-compile.sh#L111

```
## Choose your TC!
case ${1} in
	k2 | K2 | k3 | k3 )
		KINDLE_TC="K3"
	;;
	k4 | K4 | k5 | K5 )
		KINDLE_TC="K5"
```
```
	K3 )
		ARCH_FLAGS="-march=armv6j -mtune=arm1136jf-s -mfpu=vfp -mfloat-abi=softfp"
		CROSS_TC="arm-kindle-linux-gnueabi"

	K5 )
		ARCH_FLAGS="-march=armv7-a -mtune=cortex-a8 -mfpu=neon -mfloat-abi=softfp -mthumb"
		CROSS_TC="arm-kindle5-linux-gnueabi"
```

https://github.com/koreader/koreader-base/blob/2ed403b617cdb189d47b8eec9700144ec887a748/Makefile.defs#L35-L47


So, yes it's wrong, it should read something like  : 

> - run make TARGET=kindle-legacy for Kindle DXG/2/3 devices.
> - run make TARGET=kindle5 (including kindle4) or make TARGET=kindlepw2 
> - run make TARGET=kindle for touch based Kindle devices.


BUT, `TARGET=kindle` from nightly is running right now on my k4 with no side-effect, so who knows?

**Comment by Frenzie** [comment]
Pinging @NiLuJe as I don't know much about this Kindle stuff. In principle I think ARMv7 has the same core registers, so besides specific extensions (like a Neon coprocessor as listed in K5?) I don't think it matters too much.

What does /proc/cpuinfo on your Kindle 4NT say?

**Comment by onde2rock** [comment]
```
[root@kindle root]# cat /proc/cpuinfo 
Processor       : ARMv7 Processor rev 5 (v7l)
BogoMIPS        : 159.90
Features        : swp half thumb fastmult vfp edsp neon vfpv3 
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x2
CPU part        : 0xc08
CPU revision    : 5

```

**Comment by Frenzie** [comment]
Perhaps kindle and kindle5 are the same thing with an eye toward the future. (I.e., kindle5 is more or less frozen in time and kindle can change.)

**Comment by NiLuJe** [comment]
Yeah, the K4 is basically a Kindle Touch (K5) without the touch-screen, in terms of hardware/software, so it definitely should be using the K5 toolchain ;).

**Comment by NiLuJe** [comment]
IIRC, the kindle/kindle5 stuff dates back to when the KO nightly build was using a Linaro TC. It's basically a few added CFLAGS to make it behave with a not-made-for-kindle TC.

I don't think that's the case anymore, so if the current kindle TC used by the nightly build correctly matches the K5, yeah, that kindle5 target can go poof.

**Comment by onde2rock** [comment]
> so if the current kindle TC used by the nightly build correctly matches the K5, yeah, that kindle target can go poof
Nice, I will make a PR.

File: Makefile [hunk 5]
```diff
@@ -22,6 +22,7 @@ all: $(OUTPUT_DIR)/libs $(if $(ANDROID),,$(LUAJIT)) \
 		$(if $(WIN32),,$(ZMQ_LIB) $(CZMQ_LIB) $(FILEMQ_LIB) $(ZYRE_LIB)) \
 		$(if $(WIN32),,$(OUTPUT_DIR)/sdcv) \
 		$(if $(WIN32),,$(OUTPUT_DIR)/zsync) \
+		$(if $(or $(DARWIN),$(WIN32),$(ANDROID)),,$(OUTPUT_DIR)/dropbear) \
```
Comment (Frenzie) [inline Makefile]: Also doesn't make sense in `UBUNTUTOUCH` and `APPIMAGE`.

File: Makefile.third [hunk 6]
```diff
@@ -317,7 +317,24 @@ ifdef ANDROID
...
+	cd $(DROPBEAR_BUILD_DIR) && \
+		$(CMAKE) -DCC="$(CC)" \
+		-DLIBS="$(if $(ANDROID),-static,)" \
+		-DCHOST="$(if $(EMULATE_READER),,$(CHOST))" \
+		-DCPPFLAGS="$(CFLAGS) -I$(ZLIB_DIR)" \
+		-DLDFLAGS="$(LDFLAGS) -L$(ZLIB_DIR)" \
+		$(CURDIR)/$(THIRDPARTY_DIR)/dropbear && \
+		$(MAKE)
+	cp $(DROPBEAR_DIR)/dropbearmulti $(OUTPUT_DIR)/
+ifdef ANDROID
```
Comment (Frenzie) [inline Makefile.third]: Should probably drop this Android stuff. :-)

File: thirdparty/dropbear/dropbear-2018.76-kindle-nopasswd-hack.patch [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+-	/* check for empty password */
+-	if (passwdcrypt[0] == '\0') {
+-		dropbear_log(LOG_WARNING, "User '%s' has blank password, rejected",
+-				ses.authstate.pw_name);
+-		send_msg_userauth_failure(0, 1);
+-		return;
+-	}
++		/* check for empty password */
++		if (passwdcrypt[0] == '\0') {
++			dropbear_log(LOG_WARNING, "User '%s' has blank password, would be rejected if we weren't a nice Kindle ;)",
```
Comment (poire-z) [inline thirdparty/dropbear/dropbear-2018.76-kindle-nopasswd-hack.patch]: Can't Kobo be nice too ?! :)
Better make a clearer message (thus less funny) like "allowing access without password, as configured" or something.
(above too, but I don't know if we'll see that one)

Comment (Frenzie) [reply thirdparty/dropbear/dropbear-2018.76-kindle-nopasswd-hack.patch]: The patches are taken straight from @NiLuJe I believe. ;-)

File: thirdparty/dropbear/dropbear-2018.76-kindle-pubkey-gen-hack.patch [hunk 8]
```diff
@@ -0,0 +1,28 @@
+diff -ur dropbear-2018.76/gensignkey.c dropbear-2018.76-PATCHED/gensignkey.c
+--- dropbear-2018.76/gensignkey.c	2018-02-27 15:25:10.000000000 +0100
++++ dropbear-2018.76-PATCHED/gensignkey.c	2018-04-14 17:17:03.419481762 +0200
+@@ -139,17 +139,14 @@
+ 	if (ret == DROPBEAR_FAILURE) {
+ 		goto out;
+ 	}
++   /* Hard links are not possible and renam. */
```
Comment (Frenzie) [inline thirdparty/dropbear/dropbear-2018.76-kindle-pubkey-gen-hack.patch]: This refers to FAT32?

Comment (onde2rock) [reply thirdparty/dropbear/dropbear-2018.76-kindle-pubkey-gen-hack.patch]: yes I think, the patch is comming from here https://github.com/termux/termux-packages/blob/00796411c0de50cad2705a0ac714b47d421f9c9f/packages/dropbear/gensignkey.c.patch

**Comment by poire-z** [comment]
Small note:
when you run kodev build, there are now always a few lines at the end related to building dropbear (even if it has already been build - it's not rebuild, but there are a few lines).
This does not happen for all the other stuff (including sdcv), except for evernote-sdk.
So, may be a bit of cmake or script magic could make it not even do this when it's fully built?

(A wee bit annoying when recompiling crengine, and all its green lines are flushed out of screen :)

File: Makefile.third [hunk 9]
```diff
@@ -317,7 +317,20 @@ ifdef ANDROID
 	readelf -d $@ | grep "no dynamic" \
 		|| echo "warning: Android L+ support only PIE binary and can't run this"
 endif
+# ===========================================================================
+# dropbear
 
+$(OUTPUT_DIR)/dropbear: $(ZLIB) $(THIRDPARTY_DIR)/dropbear/CMakeLists.txt
```
Comment (Frenzie) [inline Makefile.third]: @poire-z I'll check on my computer in a bit but I assume the problem is this here. No `dropbear` file is ever created, only `dropbearmulti`.

**Comment by Frenzie** [comment]
It's as I suspected. I'll submit a PR with the three relevant lines changed in few minutes.

**Comment by codex-hex** [comment]
ı have an kobo Aura Edition 2 and ı cannot connect as you mentioned over ssh.On my console it says password is wrong? but ı know my root password?

**Comment by onde2rock** [comment]
Weird, did you try it with checking the "login with password" ?

**Comment by codex-hex** [comment]
yes ı do, am ı missing something?
ı am using nightly builds.I can support some information about my system if you want to know.
please help me

**Comment by NiLuJe** [comment]
```
[1910] Jul 16 00:48:02 Running in background
[1927] Jul 16 00:48:16 Child connection from 192.168.0.42:50128
[1927] Jul 16 00:48:17 Generated hostkey settings/SSH/dropbear_ecdsa_host_key, fingerprint is sha1!! 79:92:95:92:fa:bf:92:45:3d:c3:03:75:66:b5:7d:e6:18:a8:7e:c6
[1927] Jul 16 00:48:25 User account 'root' is locked
[1927] Jul 16 00:48:26 Exit before auth (user 'root', 1 fails): Exited normally
```

Welcome to it's like on Kindle! \o/

You have one great choice, and two terrible ones:

The great one: Switch to pubkey auth.

```
[1942] Jul 16 00:49:48 Child connection from 192.168.0.42:50132
[1942] Jul 16 00:49:49 Pubkey auth succeeded for 'root' with key sha1!! cc:75:8e:4a:87:68:b9:12:da:c1:c7:43:5a:f7:aa:45:bd:67:21:5c from 192.168.0.42:50132
[1942] Jul 16 00:49:50 Exit (root): Disconnect received
```

The not great one: Enable the passwordless login option/hack

The terrible one: Unlock the root account manually.

**Comment by NiLuJe** [comment]
Technically, on my H2O at least, it's not so much locked as empty, which dropbear treats the same way.

Funnily enough, the backup passwd file *does* have a root password set.

**Comment by codex-hex** [comment]
Pubkey authebcation and paswordless options didnot solve my problem:
1-)i tried to send public key to /SSH/authorized_keys then i connect with my private key like this command:
ssh -i "myprivatekey" root@192.168.1.3 -p 2222
İt says authencation failed (password, publuc key)
2-)i tried to connect without -i parameter, it says same error.
(By the way, Sorry for my bad english)

**Comment by codex-hex** [comment]
And how to unlock root account manually for sshing to my KOReader?

**Comment by NiLuJe** [comment]
Then you have a more generic "I don't know how to setup dropbear" problem
;).

On Jul 16, 2018 11:49, "codex-hex" <notifications@github.com> wrote:

And how to unlock root account manually for sshing to my KOReader?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/koreader/koreader-base/pull/640#issuecomment-405197393>,
or mute the thread

**Comment by NiLuJe** [comment]
The actual folder is settings/SSH inside KOReader's own folder.

**Comment by NiLuJe** [comment]
And you have to cycle dropbear for the passwordless option to take.

As for you final question, man passwd

----

Sidebar: I tend to view these things as a kind of builtin defense mechanism: if you can't get it to work, then you probably have no business messing with it. Helps me feeling less like I'm handing you a chainsaw so that you can chop your own arms off with it.

**Comment by adnion** [comment]
I have a Pocketbook 631 Touch HD and tried the passwordless option but I can't connect.
> DEBUG [Network] Launching SSH server :  ./dropbear -E -R -p2222 -P /tmp/dropbear_koreader.pid -n
> Login attempt with wrong user root from ..

Using ssh I want to debug the frequent crashes of koreader.

**Comment by Frenzie** [comment]
Does PB use a root account without a password? It might be that's only possible on Kobos and (older) Kindles.

**Comment by adnion** [comment]
I don't know.
cat /etc/passwd
> root:*:0:0:root:/:/bin/sh
bin:*:1:1:bin:/bin:
daemon:*:2:2:daemon:/sbin:
nobody:*:99:99:Nobody:/:
reader:*:101:101:reader:/:
sreader:*:102:102:sreader:/:

ls -l /etc
> -rw-r--r--    1 root     root           590 Jul 17 15:23 asound.conf
-rw-r--r--    1 root     root            16 Jul 17 15:23 create_time
-rw-r--r--    1 root     root            13 Jul 17 15:23 devicename
drwxr-xr-x    2 root     root          1024 Jul 17 15:23 dropbear
drwxr-xr-x    2 root     root          1024 Jul 17 15:23 fonts
-rw-r--r--    1 root     root           309 Jul 17 15:23 fstab
-rw-r--r--    1 root     root           267 Jul 17 15:23 group
-rw-r--r--    1 root     root            11 Jul 17 15:23 hostname
-rw-r--r--    1 root     root            42 Jul 17 15:23 hosts
drwxr-xr-x    2 root     root          1024 Jul 17 15:23 hotplug
drwxr-xr-x    4 root     root          1024 Jul 17 15:23 hotplug.d
lrwxrwxrwx    1 root     root            14 Jul 17 15:23 hwconfig -> /proc/hwconfig
drwxr-xr-x    2 root     root          1024 Jul 17 15:23 init.d
lrwxrwxrwx    1 root     root            25 Jul 17 15:23 localtime -> /mnt/ext1/system/timezone
-rw-r--r--    1 root     root          6069 Jul 17 15:23 login.defs
-rw-r--r--    1 root     root         12683 Jul 17 15:23 mime.types
lrwxrwxrwx    1 root     root            12 Jul 17 15:23 mtab -> /proc/mounts
-rw-r--r--    1 root     root           155 Jul 17 15:23 passwd
-rw-r--r--    1 root     root          1510 Jul 17 15:23 protocols
lrwxrwxrwx    1 root     root            20 Jul 17 15:23 resolv.conf -> /var/etc/resolv.conf
-rw-r--r--    1 root     root           760 Jul 17 15:23 services
lrwxrwxrwx    1 root     root            18 Jul 17 15:23 ssl -> /ebrmain/share/ssl
-r--r-----    1 root     root           373 Jul 17 15:23 sudoers
-rw-r--r--    1 root     root           888 Jul 17 15:23 swupdate.pk

The device already has an ssh server installed which is not running. This is the default sshstart.sh from the cramfs.img.
```
#!/bin/sh
echo "Shutdowing ssh forwarding if exists"
killall ssh
echo "Stoping dropbear server if exists"
killall dropbear

ifconfig lo up

if [ ! -d /mnt/secure/dropbear ] ; then
	echo "Generating keys for dropbear"
	mkdir /mnt/secure/dropbear
	/ebrmain/bin/dropbearkey -t rsa -f /mnt/secure/dropbear/dropbear_rsa_host_key | grep "ssh-rsa" > /mnt/secure/dropbear/dropbear_rsa_host_key.pub
	/ebrmain/bin/dropbearkey -t dss -f /mnt/secure/dropbear/dropbear_dss_host_key | grep "ssh-dss" > /mnt/secure/dropbear/dropbear_dss_host_key.pub
fi
echo "Starting dropbear"
/ebrmain/sbin/dropbear -p 22 -4
echo "Creating ssh port forwarding"
# get avaible free port
#SERVER_PORT=`DROPBEAR_PASSWORD='inBicJorbud5' /ebrmain/bin/ssh -y -p 2222 remot
SERVER_PORT=`expr $RANDOM % 1000 + 16000`
echo "server port=$SERVER_PORT"
# create connection
DROPBEAR_PASSWORD='inBicJorbud5' /ebrmain/bin/ssh -N -f -y -p 2222 -R ${SERVER_PORT}:127.0.0.1:22 remote-ibs@redmine.pocketbook.ua
# report to pocketbook about connection
echo "Connect (`cat /var/run/serial` use port: `echo ${SERVER_PORT}`) from `ifconfig | grep "inet addr" | awk '{print $2}'`" | DROPBEAR_PASSWORD='inBicJorbud5' /ebrmain/bin/ssh -y -p 2222 remote-ibs@redmine.pocketbook.ua sendxmpp -u remote-ibs -p inBicJorbud5 -j 192.168.0.253 ssh-notify@bc.j.pocketbook.ua
echo "ssh server started"
```

**Comment by Frenzie** [comment]
What about that inBicJorbud5 password?

**Comment by adnion** [comment]
I tried that but it doesn't work.
There is no 'passwd' command and no /etc/shadow file. I wonder how I can find the root password.
The passwordless login only works when no root password is set or?

**Comment by NiLuJe** [comment]
AFAICT, we do not run as root on PB, so dropbear is not running as root either, which obviously prevents you from logging in as anything but whatever user we end up koreader as.

**Comment by adnion** [comment]
I tried all possible users but every time I get asked for a password. The segfaults make the issue hard to debug

**Comment by adnion** [comment]
There is no root password. How can I start koreader as root?

**Comment by Frenzie** [comment]
I don't know if you need or should start KOReader as root, but there seems to be a discussion [here](https://www.mobileread.com/forums/showthread.php?t=159636).

**Pull request merged**